ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixBootstrapUpgradeUtil.java

```diff
@@ -245,131 +283,182 @@ private void populateInstancesAndPartitionsMap() {
           }
+        } else if (!instanceSetInStatic.equals(instanceSetInHelix)) {
+          info("Different instance sets for partition {} under resource {}. Updating Helix using static.",
+              partitionName, resourceName);
+          ArrayList<String> newInstances = new ArrayList<>(instanceSetInStatic);
+          Collections.shuffle(newInstances);
+          resourceIs.setPreferenceList(partitionName, newInstances);
+          resourceModified = true;
         }
       }
-
-      info("Done updating all existing partitions, now adding new partitions under new resources, if any.");
-      long nextResource = maxResource + 1;
-      TreeMap<String, Set<String>> partitionsUnderNextResource = new TreeMap<>();
-      for (Map.Entry<String, Set<String>> entry : partitionsToInstancesInDc.entrySet()) {
-        partitionsUnderNextResource.put(entry.getKey(), entry.getValue());
-        if (partitionsUnderNextResource.size() == maxPartitionsInOneResource) {
-          info("maxPartitions reached for resource {}, adding them under it", nextResource);
-          addPartitionsUnderResource(dcAdmin, dc.getName(), partitionsUnderNextResource, Long.toString(nextResource));
-          info("Added partitions from {} to {} under resource {}", partitionsUnderNextResource.firstKey(),
-              partitionsUnderNextResource.lastKey(), nextResource);
-          partitionsUnderNextResource.clear();
-          nextResource++;
+      if (resourceModified) {
+        if (resourceIs.getPartitionSet().isEmpty()) {
+          dcAdmin.dropResource(clusterName, resourceName);
+          resourcesDropped++;
+        } else {
+          dcAdmin.setResourceIdealState(clusterName, resourceName, resourceIs);
+          resourcesUpdated++;
         }
       }
-      if (!partitionsUnderNextResource.isEmpty()) {
-        info("last resource " + nextResource + ", actually adding all newly added partitions");
-        addPartitionsUnderResource(dcAdmin, dc.getName(), partitionsUnderNextResource, Long.toString(nextResource));
-        info("Added all remaining partitions {} to {} under resource {}", partitionsUnderNextResource.firstKey(),
-            partitionsUnderNextResource.lastKey(), nextResource);
+    }
+
+    // Add what is not already in Helix under new resources.
+    int fromIndex = 0;
+    List<Map.Entry<String, Set<String>>> newPartitions = new ArrayList<>(partitionsToInstancesInDc.entrySet());
+    while (fromIndex < newPartitions.size()) {
+      String resourceName = Integer.toString(++maxResource);
+      int toIndex = Math.min(fromIndex + maxPartitionsInOneResource, newPartitions.size());
+      List<Map.Entry<String, Set<String>>> partitionsUnderNextResource = newPartitions.subList(fromIndex, toIndex);
+      fromIndex = toIndex;
+      IdealState idealState = new IdealState(resourceName);
+      idealState.setStateModelDefRef(LeaderStandbySMD.name);
+      info("Adding partitions for instances in " + dcName);
```

Comment:
use log formatting `info("Adding partitions for instances in {}", dcName);`

Reply:
Would be better if `info("Adding partitions for next resource in {}", dcName);`  ?